### DISCUSSION TITLE
Temporal Planning

### DISCUSSION THREAD

**Discussion by mikand** [aiplan4eu/unified-planning#45] [Ideas]
Next step in WP3 is to add temporal planning features to the UPF. This is to discuss proposals and ideas for this major feature.

### Objective
Support at the very least PDDL2.1 + TILs

Additional Desiderata:
- Timed goals (e.g. [100] x == true; [end] x = false;)
- Intermediate Conditions and Effects ([start] x:= true; [start +4] x== false;)
- Duration constraints (not only fixed durations)
- Continuous change
- LTL goals? PDDL3?

### Relation with Timeline-Based Planning
We will focus on action-based temporal planning, but we have to keep in mind that in the future we want to include timeline-based planning. Timeline-based planning does not seem easily “integrable”, but we might create a dedicated `TimelineProblem` class that should re-use as much as possible from the action-based parts of the UPF. 

### Proposal
The spirit is to start from a ANML-like model with intermediate conditions and effects limited to `start+k` and `end-k`, keeping actions as intervals with duration constraints.

```python3
import upf

Location = upf.UserType('Location')
robot_at = upf.Fluent('robot_at', BoolType(), [Location])
distance = upf.Fluent('distance', RealType(), [Location, Location])

move = upf.DurativeAction('move', l_from=Location, l_to=Location)
l_from = move.parameter('l_from')
l_to = move.parameter('l_to')
move.set_duration_constraint(IntervalDuration(5, 16)) #duration >= 5 and duration <= 16
# To be decided: what to support as duration constraint/bound
# move.set_duration_constraint(FixedDuration(distance(l_from, l_to))) #duration = distance(from, to)
move.add_start_condition(Not(Equals(l_from, l_to)), delay=3) # [start + 3]
move.add_condition(StartTiming(), Not(Equals(l_from, l_to))) # [start] ...
move.add_condition(EndTiming(3), robot_at(l_from)) # [end - 3] ...
move.add_condition(EndTiming(), Not(robot_at(l_to))) # [end] ...
move.add_effect(StartTiming(), robot_at(l_from), False)
move.add_effect(EndTiming(),robot_at(l_to), True)
move.add_durative_condition(LeftOpenInterval(StartTiming(), EndTiming()), Not(robot_at(l_to))) # (all] ...

l1 = upf.Object('l1', Location)
l2 = upf.Object('l2', Location)
problem = upf.Problem('robot')
problem.add_fluent(robot_at)
problem.add_durative_action(move)
problem.add_object(l1)
problem.add_object(l2)
problem.set_initial_value(robot_at(l1), True)
problem.set_initial_value(robot_at(l2), False)
problem.add_timed_effect(100, AssignEffect(robot_at(l2), False))
problem.add_goal(robot_at(l2)) # At the end of the plan robot is in l2
problem.add_timed_goal(500, robot_at(l1)) # At time 500 robot is in l1
problem.add_maintain_goal(LeftOpenInterval(50, 100), robot_at(l1)) # Between (50 and 100] robot stays in l2
```

### Issues / Comments
- We might need more expressive constraints, such as synchronizations of timelines (e.g. [1])
- For HTN features, we need to refer to timepoints of the actions to create STN-like constraints

### Attention points
- Clarify semantics of time (discrete, epsilon, continuous)
- PDDL 2.1 uses epsilon in its semantics, we need to be clear on what happens when we parse a PDDL 2.1 problem and then use a PDDL 2.1 planner: we should not complicate the problem to change the semantics

### References
[1]  Amanda Jane Coles, Andrew Coles, Moisés Martínez, Emre Savas, Juan Manuel Delfa, Tomás de la Rosa, Yolanda E.-Martín, Angel García Olaya: Efficiently Reasoning with Interval Constraints in Forward Search Planning. AAAI 2019: 7562-7569

**Comment by hstairs** [comment]
Just a quick comment. We need to clarify whether we want to support self-overlapping or not. Self-overlapping is when multiple instances of the same action can overlap in time. Most people argue that this feature is of dubious usefulness and I tend to agree. But since PDDL2.1 in principle allows this, we need to be clear about that. Not sure what happens with ANML.

**Reply by arbimo** [reply]
There are some legitimate use cases, for instance for an action borrowing from a reservoir resource. The natural way of representing this is with a durative action that borrows at starts and gives back at the end. If several borrows occur concurrently on the same resource, then they would result in self overlapping actions. (PDDL is very impractical to model this however).

Allowing self-overlapping actions could be an opt-in feature : a boolean flag that is false by default. I don't think we need more than that.

Note that it some cases, it might be dangerous to disable by default, as it contradicts the common expectation for the purpose of allowing planners to be more effective.

**Reply by hstairs** [reply]
Ok, with the numeric planning features I can see that. I was meaning the prop fragment of temporal planning. Good point :)

Regarding the default, I am not sure that all people really expect actions to self-overlap with themselves, even if PDDL2.1 doesn't disallow it explicitly. 

We may also go for having the specification of whether we allow or disallow self-overlapping as a mandatory option to be specified.

Curiosity: How is the semantics of ANML in this regard?

**Reply by arbimo** [reply]
For ANML, the semantics are not clearly defined but I don't believe that self overlapping is disallowed. In fact, self-overlapping is not very relevant in the full ANML context as actions might have conditions/effects outside of the [start,end] interval. However I have only used this feature in high-level actions (i.e. HTN's methods) that are not explicitly distinct from primitive actions.

Another route could be to have the "requires self-overlapping" flag to true by default (at least when loading PDDL). The effect on a planner that does not support such problem is that it would be flagged as not complete. However, as it does not affect soundness, we are still free to use it, we just should not trust it when it says that no solution exists (which planners rarely do anyway :wink:). Note that we have the same incompleteness problem for, e.g., TFD or YAHSP for temporal planning, even without self-overlapping actions, so we might need a more general story to handle this.
